pkg/sdn/plugin/dns.go

```diff
@@ -0,0 +1,186 @@
+}
+
+func (d *DNS) Get(dns string) dnsValue {
+	d.lock.Lock()
+	defer d.lock.Unlock()
+
+	data := dnsValue{}
+	if res, ok := d.dnsMap[dns]; ok {
+		data.ips = make([]net.IP, len(res.ips))
+		copy(data.ips, res.ips)
+		data.ttl = res.ttl
+		data.nextQueryTime = res.nextQueryTime
+	}
+	return data
+}
+
+func (d *DNS) Add(dns string) error {
+	d.lock.Lock()
+	defer d.lock.Unlock()
+
+	if len(d.dnsMap) == 0 {
+		d.dnsMap = map[string]dnsValue{}
+	}
+	d.dnsMap[dns] = dnsValue{}
+	err, _ := d.update(dns)
+	return err
+}
+
+func (d *DNS) Update() (error, bool) {
+	d.lock.Lock()
+	defer d.lock.Unlock()
+
+	errList := []error{}
+	changed := false
+	for dns := range d.dnsMap {
+		err, updated := d.update(dns)
+		if err != nil {
+			errList = append(errList, err)
+			continue
+		}
+		if updated {
+			changed = true
+		}
+	}
+	return kerrors.NewAggregate(errList), changed
+}
+
+func (d *DNS) update(dns string) (error, bool) {
+	out, err := exec.Command("sh", []string{"-c", digShellCmdGetIPAndTTL(dns)}...).Output()
+	if err != nil || len(out) == 0 {
+		return fmt.Errorf("Failed to fetch IP addr and TTL value for domain: %q, err: %v", dns, err), false
+	}
+	outStr := strings.Trim(string(out[:]), "\n")
+
+	res, ok := d.dnsMap[dns]
+	if !ok {
+		// Should not happen, all operations on dnsMap are synchronized by d.lock
+		return fmt.Errorf("DNS value not found in dnsMap for domain: %q", dns), false
+	}
+
+	var minTTL time.Duration
+	var ips []net.IP
+	for _, line := range strings.Split(outStr, "\n") {
+		fields := strings.Fields(line)
+		if len(fields) != 5 {
+			continue
+		}
+
+		ttl, err := time.ParseDuration(fmt.Sprintf("%ss", fields[1]))
+		if err != nil {
+			glog.Errorf("Invalid TTL value for domain: %q, err: %v, defaulting ttl=%s", dns, err, defaultTTL.String())
+			ttl = defaultTTL
+		}
+		if (minTTL.Seconds() == 0) || (minTTL.Seconds() > ttl.Seconds()) {
+			minTTL = ttl
+		}
+
+		ip := net.ParseIP(fields[4])
+		if ip != nil {
+			ips = append(ips, ip)
+		}
+	}
+
+	changed := false
+	if !equal(res.ips, ips) {
+		changed = true
+	}
+	res.ips = ips
+	res.ttl = minTTL
+	res.nextQueryTime = time.Now().Add(res.ttl)
+	d.dnsMap[dns] = res
+	return nil, changed
+}
+
+func (d *DNS) GetMinQueryTime() (time.Time, bool) {
+	d.lock.Lock()
+	defer d.lock.Unlock()
+
+	timeSet := false
+	var minTime time.Time
+	for _, res := range d.dnsMap {
+		if (timeSet == false) || res.nextQueryTime.Before(minTime) {
+			timeSet = true
+			minTime = res.nextQueryTime
+		}
+	}
+
+	return minTime, timeSet
+}
+
+// Due to lack of any go bindings for dns resolver that actually provides TTL value, we are relying on 'dig' shell command.
+
+func digShellCmdGetIPAndTTL(dns string) string {
+	// Output Format:
+	// <domain-name>.		<<ttl from authoritative ns>	IN	A	<IP addr>
+	return fmt.Sprintf("%s +nocmd +noall +answer +ttlid a %s", DIG, dns)
+}
+
+func equal(oldips, newips []net.IP) bool {
+	if len(oldips) != len(newips) {
+		return false
+	}
+
+	for _, oldip := range oldips {
```

Comment:
An alternative would be to define an "ipsArray" type that implemented Sort (eg, Len(), Less(), and Swap()), sort at the end and do a direct compare.  Possibly more idiomatic but eh.